File: usecases/src/main/java/com/zeyad/usecases/stores/DataStore.java
package com.zeyad.usecases.stores;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.gson.Gson;
import com.zeyad.usecases.Config;
import com.zeyad.usecases.db.RealmQueryProvider;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.util.HashMap;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;

/**
 * Interface that represents a data store from where data is retrieved.
 */
public interface DataStore {
    Gson gson = Config.INSTANCE.getGson();

    @NonNull
    <M> Flowable<List<M>> dynamicGetList(String url, String idColumnName, Class requestType,
                                         boolean persist, boolean shouldCache);

    /**
     * Get an {@link Flowable} which will emit a Object by its id.
     */
    @NonNull
    <M> Flowable<M> dynamicGetObject(String url, String idColumnName, Object itemId, Class itemIdType,
                                     Class requestType, boolean persist, boolean shouldCache);

    /**
     * Search disk with a RealmQuery which returns an {@link Flowable} that will emit a list of
     * Object.
     */
    @NonNull
    <M> Flowable<List<M>> queryDisk(RealmQueryProvider queryFactory);

    /**
     * Patch a JSONObject which returns an {@link Flowable} that will emit a Object.
     */
    @NonNull
    <M> Flowable<M> dynamicPatchObject(String url, String idColumnName, Class itemIdType,
                                       @NonNull JSONObject jsonObject, Class requestType,
                                       Class<M> responseType, boolean persist, boolean cache,
                                       boolean queuable);

    /**
     * Post a JSONObject which returns an {@link Flowable} that will emit a Object.
     */
    @NonNull
    <M> Flowable<M> dynamicPostObject(String url, String idColumnName, Class itemIdType,
                                      JSONObject keyValuePairs, Class requestType, Class<M> responseType,
                                      boolean persist, boolean cache, boolean queuable);

    /**
     * Post a HashMap<String, Object> which returns an {@link Flowable} that will emit a list of
     * Object.
     */
    @NonNull
    <M> Flowable<M> dynamicPostList(String url, String idColumnName, Class itemIdType,
                                    JSONArray jsonArray, Class requestType, Class<M> responseType,
                                    boolean persist, boolean cache, boolean queuable);

    /**
     * Put a HashMap<String, Object> disk with a RealmQuery which returns an {@link Flowable} that
     * will emit a Object.
     */
    @NonNull
    <M> Flowable<M> dynamicPutObject(String url, String idColumnName, Class itemIdType,
                                     JSONObject keyValuePairs, Class requestType, Class<M> responseType,
                                     boolean persist, boolean cache, boolean queuable);

    /**
     * Put a HashMap<String, Object> disk with a RealmQuery which returns an {@link Flowable} that
     * will emit a list of Object.
     */
    @NonNull
    <M> Flowable<M> dynamicPutList(String url, String idColumnName, Class itemIdType,
                                   JSONArray jsonArray, Class requestType, Class<M> responseType,
                                   boolean persist, boolean cache, boolean queuable);

    /**
     * Delete a HashMap<String, Object> from cloud which returns an {@link Flowable} that will emit
     * a Object.
     */
    @NonNull
    <M> Flowable<M> dynamicDeleteCollection(String url, String idColumnName, Class itemIdType,
                                            JSONArray jsonArray, Class requestType, Class<M> responseType,
                                            boolean persist, boolean cache, boolean queuable);

    /**
     * Delete all items of the same type from cloud or disk which returns an {@link Completable}
     * that will emit a list of Object.
     */
    @NonNull
    Single<Boolean> dynamicDeleteAll(Class requestType);

    @NonNull
    Flowable<File> dynamicDownloadFile(String url, File file, boolean onWifi, boolean whileCharging,
                                       boolean queuable);

    @NonNull
    <M> Flowable<M> dynamicUploadFile(String url, @NonNull HashMap<String, File> keyFileMap,
                                      @Nullable HashMap<String, Object> parameters, boolean onWifi,
                                      boolean whileCharging, boolean queuable, @NonNull Class<M> responseType);
}

File: usecases/src/main/java/com/zeyad/usecases/stores/DataStore.kt
package com.zeyad.usecases.stores

import com.zeyad.usecases.db.RealmQueryProvider
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import io.realm.RealmModel
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.*

/**
 * Interface that represents a data store from where data is retrieved.
 */
interface DataStore {

    fun <M> dynamicGetList(url: String, idColumnName: String, requestType: Class<M>,
                           persist: Boolean, shouldCache: Boolean): Flowable<List<M>>

    /**
     * Get an [Flowable] which will emit a Object by its id.
     */
    fun <M> dynamicGetObject(url: String, idColumnName: String, itemId: Any, itemIdType: Class<*>,
                             requestType: Class<M>, persist: Boolean, shouldCache: Boolean): Flowable<M>

    /**
     * Search disk with a RealmQuery which returns an [Flowable] that will emit a list of
     * Object.
     */
    fun <M : RealmModel> queryDisk(queryFactory: RealmQueryProvider<M>): Flowable<List<M>>

    /**
     * Patch a JSONObject which returns an [Flowable] that will emit a Object.
     */
    fun <M> dynamicPatchObject(url: String, idColumnName: String, itemIdType: Class<*>,
                               jsonObject: JSONObject, requestType: Class<*>,
                               responseType: Class<M>, persist: Boolean, cache: Boolean,
                               queuable: Boolean): Flowable<M>

    /**
     * Post a JSONObject which returns an [Flowable] that will emit a Object.
     */
    fun <M> dynamicPostObject(url: String, idColumnName: String, itemIdType: Class<*>,
                              jsonObject: JSONObject, requestType: Class<*>, responseType: Class<M>,
                              persist: Boolean, cache: Boolean, queuable: Boolean): Flowable<M>

    /**
     * Post a HashMap<String></String>, Object> which returns an [Flowable] that will emit a list of
     * Object.
     */
    fun <M> dynamicPostList(url: String, idColumnName: String, itemIdType: Class<*>,
                            jsonArray: JSONArray, requestType: Class<*>, responseType: Class<M>,
                            persist: Boolean, cache: Boolean, queuable: Boolean): Flowable<M>

    /**
     * Put a HashMap<String></String>, Object> disk with a RealmQuery which returns an [Flowable] that
     * will emit a Object.
     */
    fun <M> dynamicPutObject(url: String, idColumnName: String, itemIdType: Class<*>,
                             jsonObject: JSONObject, requestType: Class<*>, responseType: Class<M>,
                             persist: Boolean, cache: Boolean, queuable: Boolean): Flowable<M>

    /**
     * Put a HashMap<String></String>, Object> disk with a RealmQuery which returns an [Flowable] that
     * will emit a list of Object.
     */
    fun <M> dynamicPutList(url: String, idColumnName: String, itemIdType: Class<*>,
                           jsonArray: JSONArray, requestType: Class<*>, responseType: Class<M>,
                           persist: Boolean, cache: Boolean, queuable: Boolean): Flowable<M>

    /**
     * Delete a HashMap<String></String>, Object> from cloud which returns an [Flowable] that will emit
     * a Object.
     */
    fun <M> dynamicDeleteCollection(url: String, idColumnName: String, itemIdType: Class<*>,
                                    jsonArray: JSONArray, requestType: Class<*>, responseType: Class<M>,
                                    persist: Boolean, cache: Boolean, queuable: Boolean): Flowable<M>

    /**
     * Delete all items of the same type from cloud or disk which returns an [Completable]
     * that will emit a list of Object.
     */
    fun dynamicDeleteAll(requestType: Class<*>): Single<Boolean>

    fun dynamicDownloadFile(url: String, file: File, onWifi: Boolean, whileCharging: Boolean,
                            queuable: Boolean): Flowable<File>

    fun <M> dynamicUploadFile(url: String, keyFileMap: HashMap<String, File>, parameters: HashMap<String, Any>,
                              onWifi: Boolean, whileCharging: Boolean, queuable: Boolean,
                              responseType: Class<M>): Flowable<M>
}
